### ISSUE TITLE
Change init_threadpool_tbb() to accept number of threads and deprecate STAN_NUM_THREADS

### ISSUE THREAD

**Issue by rok-cesnovar** [stan-dev/math#1949]
## Description

This is the math issue for https://github.com/stan-dev/cmdstan/issues/825

`init_threadpool_tbb()` should accept an integer argument that will set the maximum number of threads. The max number of threads is currently read from the `STAN_NUM_THREADS` environment variable, which we decided should be deprecated in favor of this new interface. 

This will enable setting the number of threads as an argument in cmdstan and other interfaces as well hopefully.

#### Current Version:
v3.2.0

**Comment by rok-cesnovar** [comment]
I have this mostly done but got stuck with testing due to the way the threadpool initalizes/terminates. Its fine how it works it just may not be convenient for testing.

Example (see last line):
```cpp
  tbb::task_scheduler_init& tbb_init = stan::math::init_threadpool_tbb(10);
  EXPECT_TRUE(tbb_init.is_active());

  EXPECT_EQ(10, tbb::this_task_arena::max_concurrency());

  tbb_init.terminate();
  EXPECT_FALSE(tbb_init.is_active());

  tbb::task_scheduler_init& tbb_init_2 = stan::math::init_threadpool_tbb(4);
  EXPECT_TRUE(tbb_init_2.is_active()); // this is false
```
@wds15 Is there a way to terminate and reinitialize the threadpool within the same executable? I want to add similar tests that we have right now. If not, I will just make a few more test files. Not a huge problem, just inconvenient.

**Comment by rok-cesnovar** [comment]
I played around with the task_arena Sebastian suggested, but then the `tbb::this_task_arena::max_concurrency()` does not return the proper values (always the number of cores).

I think it would be better to postpone this one for the next release. I dont feel comfortable with changing this now at the last second.

**Comment by SteveBronder** [comment]
Agree on waiting. We should do this for OpenCL as well

**Comment by rok-cesnovar** [comment]
I was unable to make this work. tbb::this_task_arena::max_concurrency() always returns the number of cores, so I am unassigning myself from this. If anyone else has time and ideas for this that would be very helpful. Still think this issue should be done.

**Comment by wds15** [comment]
You have to run things in a custom (speak: newly created) task_arena. So we either create a global stan-math task_arena or run stuff in `task_arena`s which we create when firing up the sampler. I would much prefer the non-global route. Once you have a task_arena created, you can then dispatch work to it which will run with the maximal concurrency as specified when instantiating the task_arena.

**Comment by rok-cesnovar** [comment]
Yeah, that is the only way probably. But that requires changes above the Math library. Meaning that all reduce_sum and map_rect calls will require an arena argument right? It would be the proper way, that I agree.

**Comment by wds15** [comment]
Yes, math does not need to deal with the arena stuff. That is something which should to go into the services layer.

**Comment by hsbadr** [comment]
> I played around with the task_arena Sebastian suggested, but then the `tbb::this_task_arena::max_concurrency()` does not return the proper values (always the number of cores).

`tbb::this_task_arena::max_concurrency()` should return the number of cores. You need to get `max_concurrency()` of the initialized arena (`tbb::task_arena& tbb_init = stan::math::init_threadpool_tbb();`): `tbb_init.max_concurrency()`

**Comment by rok-cesnovar** [comment]
And I think there lies the problem. When running the actual reduce_sum we can only access tbb::this_task_arena and not the initialized one. We would need a to create a global task arena.

**Comment by wds15** [comment]
Yes. I would think that the services should run the chains in a task_arena which they control themselves rather than having a global task_arena. This needs quite a bit of plumbing in the services which have never been threading aware (which is why we started with the environment variable approach).

**Comment by hsbadr** [comment]
@wds15 That'd be tricky in terms of the available vs hardware threads, especially if those services can run simultaneously. I need to learn more about the services you're talking about before suggesting a solution.

**Comment by hsbadr** [comment]
@rok-cesnovar Generally speaking, we could create an instance of `task_arena` for each service and attach it to the internal arena currently used by the calling thread, via `tbb::task_arena(tbb::task_arena::attach())`, together with a global `task_scheduler_observer`.
